superset/key_value/utils.py

```diff
@@ -63,3 +64,9 @@ def get_uuid_namespace(seed: str) -> UUID:
     md5_obj = md5()
     md5_obj.update(seed.encode("utf-8"))
     return UUID(md5_obj.hexdigest())
+
+
+def get_deterministic_uuid(namespace: str, payload: Any) -> UUID:
+    """Get a deterministic UUID (uuid3) from a salt and a payload."""
+    payload_str = json_dumps_w_dates(payload, sort_keys=True)
```

Comment:
@villebro @michael-s-molina any particular reason why KeyValue was stored with Python pickle instead of JSON? IMO we should avoid storing pickles in database as they are not portable and could be versioned out if an advanced python object was used. It'd also be impossible to export the database directly for offline analysis.

Reply:
@ktmud not all `KeyValue` implementations will necessarily be JSON serializable. For instance, the `SupersetMetastoreCache` KeyValue objects aren't JSON serialisable. While pickle isn't necessarily perfect in all scenarios, IMO it's the most generally supported serialising/deserializing utility for Python, and as such seems like the best solution for a general purpose key-value store.

At the time this feature was designed there was really no requirement for analysing the contents of the value contents, so this wasn't considered. But if this is needed I don't see any reason why we couldn't make the serializer and deserialiser customizable to make it possible to use another library for encoding/decoding the values and then swapping it out for `json.dumps` and `json.loads` for the permalink implementations (would of course require a migration).

In the case of the proposed function, I'd recommend adding a comment in the doctoring that the payload needs to be JSON serializable.

Reply:
I still don't think it's a good idea to store language-specific serialized data in databases---even though Superset is a predominant Python app. I can see the case when we need to store other binary data, e.g., image files, cache for zipped exporst, encrypted data, etc, so maybe the `value` column indeed needs to be binary. A customizable serializer sounds like a good idea. Let's revisit when offline analysis becomes a real need.